### PR TITLE
feat: add slug property to app, use in URLs

### LINKED ISSUE #4571: Non URL-safe characters in app names cause them to be inaccessible over subdomains
# Description
When using coder_apps with `subdomain=true` every time you try to open the app using the button in the ws dashboard coder shows a `404 - Application not found` page.

# Steps to reproduce
1. Define a coder app in your main.tf like this:
```

resource "coder_app" "codeserver" {
  agent_id  = coder_agent.dev.id
  name      = "Code Server"
  url       = "http://localhost:8080/?folder=/home/coder/workspace"
  icon      = "/icon/code.svg"
  subdomain = true
  healthcheck {
    url	  = "http://localhost:8080/healthz"
    interval  = 2
    threshold = 10
  }
}
```
2. Push the template change
3. Start the workspace
4. Try to open codeserver

# Addition Infos
Coder Version: [Coder v0.9.10+574e5d3](https://github.com/coder/coder/commit/574e5d37c76cb0b154ad62ccd4582e588235b15a)

Coder has a `CODER_WILDCARD_ACCESS_URL` configured to `*.coder.example.com`

### LINKED ISSUE #4674: Possible issue with subdomains
I keep getting 404's when trying to use the `subdomain = true` option in the Terraform config.

My configuration is as below:

```
resource "coder_app" "pgadmin" {
  agent_id = coder_agent.main.id
  name     = "PGAdmin"
  url      = "http://localhost:8800/"
  icon     = "https://simpleicons.org/icons/postgres.svg"
  subdomain = true
}
```

Which generates the subdomain `https://pgadmin--main--workspacename--username.coder.domain/` in the workspace link.

However when I go to this link, the call does not get proxied down into the container. If I curl `http://localhost:8800` in the container it works fine. 

Can you please advise if I am doing something wrong?

**EDIT**
I have just tried using `https://8800--main--workspacename--username.coder.domain/` and this works so the service is reachable but the coder_app subdomain stuff is not proxying correctly

### PR DESCRIPTION
**Pull request by deansheather**
Closes #4571, Closes #4674

Adds `slug` property to `workspace_apps` in the database, which is copied from the new required terraform `slug` property in templates. This new slug property is the only property accepted in application URLs including proxy path URLs. Existing bookmarks or tabs will need to be recreated from the dashboard to get new URLs in most cases (unless the app's `display_name` field matches the `slug` field).

The app "display_name" field is still used for the button in the dashboard.

`coder_app` resource names must match the regex `provisioner.ValidAppNameRegex`:
```
^[a-z0-9](-?[a-z0-9])*$
```

Which is essentially:
1. Must start with alphanumeric
2. Can contain alphanumeric or hyphen characters
3. Cannot contain two consecutive hyphens
4. Must end with alphanumeric

image

## TODO:
- [x] create new slug field in the terraform provider
- [x] add a test that ensures invalid workspace app slugs get rejected by the terraform provisioner
- [x] update the terraform provider in all templates

### CODE REVIEW

File: site/src/components/AppLink/AppLink.stories.tsx [hunk 1]
```diff
@@ -13,7 +13,8 @@ export const WithIcon = Template.bind({})
 WithIcon.args = {
   username: "developer",
   workspaceName: MockWorkspace.name,
-  appName: "code-server",
+  appSlug: "code-server",
+  appName: "Code Server",
```
Comment (jsjoeio) [inline site/src/components/AppLink/AppLink.stories.tsx:17]: nit: we always spell it "code-server"

File: site/src/components/AppLink/AppLink.tsx [hunk 2]
```diff
@@ -43,19 +45,25 @@ export const AppLink: FC<PropsWithChildren<AppLinkProps>> = ({
   health,
 }) => {
   const styles = useStyles()
+  if (appSlug === "") {
+    appSlug = appName
+  }
+  if (appName === "") {
+    appName = appSlug
+  }
```
Comment (jsjoeio) [inline site/src/components/AppLink/AppLink.tsx:53]: ❓What if both `appSlug` and `appName` are empty strings? Is that something we should handle?

**Review by jsjoeio — APPROVED** [review]
FE: great work!

File: site/src/components/AppLink/AppLink.tsx (hunk 2 above)
Comment (deansheather) [reply site/src/components/AppLink/AppLink.tsx:53]: On the backend this is impossible, one of them will definitely be set.

**Review by coadler — APPROVED** [review]

File: site/src/components/AppLink/AppLink.stories.tsx [hunk 3]
```diff
@@ -41,14 +44,16 @@ export const HealthInitializing = Template.bind({})
...
   health: "initializing",
 }
 
 export const HealthUnhealthy = Template.bind({})
 HealthUnhealthy.args = {
   username: "developer",
   workspaceName: MockWorkspace.name,
-  appName: "code-server",
+  appSlug: "code-server",
+  appName: "Code Server",
```
Comment (kylecarbs) [inline site/src/components/AppLink/AppLink.stories.tsx:57]: I think it'd be better to call this `appDisplayName` instead. Including in the Terraform, it'd be best to deprecate `name` for `display_name`, and require `slug`.

File: coderd/database/migrations/000066_app_display_name.down.sql [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+-- they would then be renamed to e.g. asdf, asdf1234, and asdf5678.
+UPDATE
+    workspace_apps
+SET
+	display_name = workspace_apps.display_name || floor(random() * 10000)::text
+FROM
+    row_numbers
+WHERE
+    workspace_apps.id = row_numbers.id AND
+    row_numbers.row_number > 1;
```
Comment (coadler) [inline coderd/database/migrations/000066_app_display_name.down.sql:26]: seems to be using spaces and tabs

File: coderd/database/migrations/000065_app_slug.up.sql [hunk 5]
```diff
@@ -7,3 +9,8 @@ UPDATE "workspace_apps" SET "slug" = "name";
 -- make "slug" column not nullable and remove default
 ALTER TABLE "workspace_apps" ALTER COLUMN "slug" SET NOT NULL;
 ALTER TABLE "workspace_apps" ALTER COLUMN "slug" DROP DEFAULT;
+
+-- add unique index on "slug" column
+ALTER TABLE "workspace_apps" ADD CONSTRAINT "workspace_apps_agent_id_slug_key" UNIQUE ("agent_id", "slug");
```
Comment (coadler) [inline coderd/database/migrations/000065_app_slug.up.sql:14]:
```suggestion
ALTER TABLE "workspace_apps" ADD CONSTRAINT "workspace_apps_agent_id_slug_idx" UNIQUE ("agent_id", "slug");
```

File: coderd/database/migrations/000066_app_display_name.up.sql [hunk 6]
```diff
@@ -0,0 +1,9 @@
+BEGIN;
```
Comment (coadler) [inline coderd/database/migrations/000066_app_display_name.up.sql:1]: Is there a purpose to having these in separate migrations? I think we'll need to bump these too

**Review by coadler — APPROVED** [review]

File: coderd/database/migrations/000065_app_slug.up.sql (hunk 5 above)
Comment (deansheather) [reply coderd/database/migrations/000065_app_slug.up.sql:14]: wait nevermind I'm thinking of the wrong file

File: coderd/database/migrations/000066_app_display_name.up.sql (hunk 6 above)
Comment (deansheather) [reply coderd/database/migrations/000066_app_display_name.up.sql:1]: not really that's just the way they were made, they're kinda large though so I'm just gonna keep them as separate files

File: coderd/database/migrations/000065_app_slug.up.sql (hunk 5 above)
Comment (deansheather) [reply coderd/database/migrations/000065_app_slug.up.sql:14]: fixed

**Pull request merged**